Source/Scene/DebugInspector.js

```diff
@@ -0,0 +1,155 @@
+import clone from "../Core/clone.js";
+import Color from "../Core/Color.js";
+import DrawCommand from "../Renderer/DrawCommand.js";
+import ShaderSource from "../Renderer/ShaderSource.js";
+import ShaderProgram from "../Renderer/ShaderProgram.js";
+import defined from "../Core/defined.js";
+
+/**
+ * @private
+ */
+function DebugInspector() {
+  this._cachedShowFrustumsShaders = {};
+}
+
+function getAttributeLocations(shaderProgram) {
+  var attributeLocations = {};
+  var attributes = shaderProgram.vertexAttributes;
+  for (var a in attributes) {
+    if (attributes.hasOwnProperty(a)) {
+      attributeLocations[a] = attributes[a].index;
+    }
+  }
+
+  return attributeLocations;
+}
+
+function createDebugShowFrustumsShaderProgram(scene, shaderProgram) {
+  var context = scene.context;
+  var sp = shaderProgram;
+  var fs = sp.fragmentShaderSource.clone();
+
+  var targets = [];
+  fs.sources = fs.sources.map(function (source) {
+    source = ShaderSource.replaceMain(source, "czm_Debug_main");
+    var re = /gl_FragData\[(\d+)\]/g;
+    var match;
+    while ((match = re.exec(source)) !== null) {
+      if (targets.indexOf(match[1]) === -1) {
+        targets.push(match[1]);
+      }
+    }
+    return source;
+  });
+  var length = targets.length;
+
+  var newMain = "";
+  newMain += "uniform vec3 debugShowCommandsColor;\n";
+  newMain += "uniform vec3 debugShowFrustumsColor;\n";
+  newMain += "void main() \n" + "{ \n" + "    czm_Debug_main(); \n";
+
+  // set debugShowCommandsColor to Color(1.0, 1.0, 1.0, 1.0) to stop rendering scene.debugShowCommands
+  // set debugShowFrustumsColor to Color(1.0, 1.0, 1.0, 1.0) to stop rendering scene.debugShowFrustums
+  var i;
+  if (length > 0) {
+    for (i = 0; i < length; ++i) {
+      newMain +=
+        "    gl_FragData[" + targets[i] + "].rgb *= debugShowCommandsColor;\n";
+      newMain +=
+        "    gl_FragData[" + targets[i] + "].rgb *= debugShowFrustumsColor;\n";
+    }
+  } else {
+    newMain += "    gl_FragColor.rgb *= debugShowCommandsColor;\n";
+    newMain += "    gl_FragColor.rgb *= debugShowFrustumsColor;\n";
+  }
+  newMain += "}";
+
+  fs.sources.push(newMain);
+
+  var attributeLocations = getAttributeLocations(sp);
+
+  return ShaderProgram.fromCache({
+    context: context,
+    vertexShaderSource: sp.vertexShaderSource,
+    fragmentShaderSource: fs,
+    attributeLocations: attributeLocations,
+  });
+}
+
+var scratchFrustumColor = new Color();
+function createDebugShowFrustumsUniformMap(scene, command) {
+  // setup uniform for the shader
+  var debugUniformMap;
+  if (!defined(command.uniformMap)) {
+    debugUniformMap = {};
+  } else {
+    debugUniformMap = clone(command.uniformMap);
+    // debugUniformMap = command.uniformMap;
+  }
+
+  if (
+    defined(debugUniformMap.debugShowFrustumsColor) ||
+    defined(debugUniformMap.debugShowFrustumsColor)
+  ) {
+    return debugUniformMap;
+  }
```

Comment:
I don't think this condition will actually get hit because `debugUniformMap` is always cloned from the original which doesn't contain these properties.

Should `clone(command.uniformMap);` be `command.uniformMap` after all? I see it was commented out.

Reply:
yup this check is initially used for the case `debugUniformMap = command.uniformMap`. I temporarily left it here since I get a bug when using the shallow copy for uniformMap (I documented the bug in the reply to your comment above)